osquery/tables/system/hash.cpp

```diff
@@ -139,7 +158,108 @@ std::string hashFromFile(HashType hash_type, const std::string& path) {
   }
 }
 
-namespace tables {
+/*
+ * FileHashCache implements persistent in-memory caching for files' hashes.
+ * this cache has LRU eviction policy. the hash is recalculated
+ * every time the mtime or size of the file changes.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#if defined(WIN32)
+
+typedef struct _stat Stat;
+#define xstat(path, st) _stat((path), static_cast<struct _stat *>(st))
+
+#else
+
+#include <unistd.h>
+
+typedef struct stat Stat;
+#define xstat(path, st) stat((path), static_cast<struct stat *>(st))
+
+#endif
+
+struct FileHashCache {
+  time_t mtime; // last modification time of the file, # sec since UNIX
+  time_t cache_access; // last access to the cache entry, # sec since UNIX
+  off_t size; // file size in bytes
+  MultiHashes hashes; // payload
+  std::string path; // tracking it for eviction policy
+
+  // comparison function for organizing the LRU heap
+  static bool greater(const FileHashCache* l, const FileHashCache* r) {
+    return l->cache_access > r->cache_access;
+  }
+  // do-it-all access function: maintains the cache of Row objects,
+  // stats file at path, if it has changed or it is not present in cache
+  // calculates the hashes and caches the result
+  // returns true if succeded, false if something went wrong
+  static bool load(const std::string& path, MultiHashes& out);
+};
+
+bool FileHashCache::load(const std::string& path, MultiHashes& out) {
+  static std::mutex mx; // synchronize the access to cache
```

Comment:
For this and your `lock_guard` below, we abstract some `mutex` helper functions in [the core header](https://github.com/facebook/osquery/blob/master/include/osquery/core.h#L203), feel free to make use of those.

Reply:
Didn't find any function that would be helpful in this case, but changed the definitions to the aliases to boost stuff defined there.